File: java/src/main/java/com/ywh/sorting/SelectionSort.java
package com.ywh.sorting;

/**
 * 选择排序
 * [排序] [不稳定排序]
 *
 * @author ywh
 * @since 13/11/2019
 */
public class SelectionSort {

    private void swap(int[] arr, int i, int j) {
        int tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }

    /**
     * Time: O(n^2), Space: O(1)
     *
     * @param arr
     */
    public void sort(int[] arr) {
        if (arr == null || arr.length == 0) {
            return;
        }
        for (int i = 0; i < arr.length; i++) {
            int minIdx = i;
            for (int j = i + 1; j < arr.length; j++) {
                if (arr[j] < arr[minIdx]) {
                    minIdx = j;
                }
            }
            swap(arr, i, minIdx);
        }
    }

    public void sortFromEnd(int[] arr) {
        if (arr == null || arr.length == 0) {
            return;
        }
        for (int i = arr.length - 1; i > 0; --i) {
            int maxIdx = i;
            for (int j = 0; j < i; j++) {
                if (arr[j] > arr[maxIdx]) {
                    maxIdx = j;
                }
            }
            swap(arr, i, maxIdx);
        }
    }
}

File: python/src/algorithm/sorting/selection_sort.py
# coding=utf-8
from typing import List


class SelectionSort(object):

    def sort(self, arr: List[int]):
        """

        :param arr:
        :return:
        """
        if not arr:
            return
        for i in range(len(arr)):
            min_idx = i
            for j in range(i + 1, len(arr)):
                if arr[j] < arr[min_idx]:
                    min_idx = j
            arr[min_idx], arr[i] = arr[i], arr[min_idx]

    def sort2(self, arr: List[int]):
        """

        :param arr:
        :return:
        """
        if not arr:
            return
        for i in range(len(arr) - 1, 0, -1):
            max_idx = i
            for j in range(0, i):
                if arr[j] > arr[max_idx]:
                    max_idx = j
            arr[max_idx], arr[i] = arr[i], arr[max_idx]